sycl/plugins/esimd_emulator/pi_esimd_emulator.hpp

```diff
@@ -104,48 +109,82 @@ struct _pi_mem : _pi_object {
 
   pi_context Context;
 
+  // To be used for piEnqueueMemBufferMap
   char *MapHostPtr = nullptr;
 
-  // Mutex for load/store accessing
+  // Mutex for load/store accessing to CM-managed surfaces from memory
+  // instrinsic implementations
   std::mutex mutexLock;
 
-  // Surface index used by CM
-  int SurfaceIndex;
+  // Surface index
+  unsigned int SurfaceIndex;
 
   virtual ~_pi_mem() = default;
 
   _pi_mem_type getMemType() const { return MemType; };
 
+  // TODO: Runtime guarantees LIFO across multiple threads for
+  // piEnqueueMemBufferMap/Unmap? No duplicated mappings for same
+  // (base, offset)?
+  std::stack<void *> Mappings;
+
 protected:
   _pi_mem(pi_context ctxt, char *HostPtr, _pi_mem_type MemTypeArg,
-          int SurfaceIdxArg)
+          unsigned int SurfaceIdxArg)
       : Context{ctxt}, MapHostPtr{HostPtr},
         SurfaceIndex{SurfaceIdxArg}, MemType{MemTypeArg} {}
 
 private:
   _pi_mem_type MemType;
 };
 
+struct cm_buffer_ptr_slot {
+  // 'UP' means 'User-Provided' in CM Lib - corresponding to
+  // Host-created buffer in SYCL
+
+  enum type { type_none, type_regular, type_user_provided };
+  type tag = type_none;
+
+  union {
+    cm_support::CmBuffer *RegularBufPtr = nullptr;
+    cm_support::CmBufferUP *UPBufPtr;
+  };
+};
+
+struct cm_image_ptr_slot {
```

Comment:
it seems these two (cm_image_ptr_slot and cm_buffer_ptr_slot) can be merged, and both enum and union extended to 4 elements. less code duplication